File: suite/Takl/Takl.ml
exception EmptyList

let null ls =
  match ls with 
    | [] -> true
    | _ -> false

let tail l = 
  match l with
    | [] -> raise EmptyList
    | _::xs -> xs

let rec len l = 
  match l with
    | [] -> 0
    | _::xs -> 1 + (len xs)

let rec list_n n = 
  if n=0 then [] else n :: (list_n (n-1))

let rec shorterp x y = 
  if null y then false 
  else if null x then true 
  else shorterp (tail x) (tail y)

let rec mas x y z = 
  if not (shorterp y x) then z 
  else 
    mas 
      (mas (tail x) y z)
      (mas (tail y) z x)
      (mas (tail z) x y)

let rec main_loop iters x y z = 
  let res = len (mas (list_n x) (list_n y) (list_n z)) in 
  if iters = 1 then 
    print_endline (string_of_int res)
  else main_loop (iters-1) x y z 

let main = 
  let iters = int_of_string Sys.argv.(1) in 
  let x = int_of_string Sys.argv.(2) in 
  let y = int_of_string Sys.argv.(3) in 
  let z = int_of_string Sys.argv.(4) in 
  main_loop iters x y z 

File: suite/Takl/Takl.rs
use std::rc::Rc;

#[derive(Clone)]
enum List<A> {
    Nil,
    Cons(A, Rc<List<A>>),
}

impl<A> List<A> {
    fn null(&self) -> bool {
        match self {
            List::Nil => true,
            List::Cons(_, _) => false,
        }
    }

    fn tail(self) -> List<A>
    where
        A: Clone,
    {
        match self {
            List::Nil => panic!("Cannot take tail of empty list"),
            List::Cons(_, as_) => Rc::unwrap_or_clone(as_),
        }
    }

    fn tail_ref<'a>(&'a self) -> &'a List<A> {
        match self {
            List::Nil => panic!("Cannot take tail of empty list"),
            List::Cons(_, as_) => &(*as_),
        }
    }

    fn len(&self) -> usize {
        match self {
            List::Nil => 0,
            List::Cons(_, as_) => 1 + as_.len(),
        }
    }

    fn shorterp<B>(&self, other: &List<B>) -> bool {
        if other.null() {
            false
        } else if self.null() {
            true
        } else {
            self.tail_ref().shorterp(other.tail_ref())
        }
    }
}

impl List<u64> {
    fn n(n: u64) -> List<u64> {
        if n == 0 {
            List::Nil
        } else {
            List::Cons(n, Rc::new(List::n(n - 1)))
        }
    }
}

fn mas(x: List<u64>, y: List<u64>, z: List<u64>) -> List<u64> {
    if !y.shorterp(&x) {
        z
    } else {
        mas(
            mas(x.clone().tail(), y.clone(), z.clone()),
            mas(y.clone().tail(), z.clone(), x.clone()),
            mas(z.tail(), x, y),
        )
    }
}

fn main_loop(iters: u64, x: u64, y: u64, z: u64) {
    let res = mas(List::n(x), List::n(y), List::n(z)).len();
    if iters == 1 {
        println!("{res}");
    } else {
        main_loop(iters - 1, x, y, z)
    }
}

fn main() {
    let mut args = std::env::args();
    args.next();
    let iters = args
        .next()
        .expect("Missing Argument iterations")
        .parse::<u64>()
        .expect("Iterations must be a number");
    let x = args
        .next()
        .expect("Missing Argument x")
        .parse::<u64>()
        .expect("x must be a number");
    let y = args
        .next()
        .expect("Missing Argument y")
        .parse::<u64>()
        .expect("y must be a number");
    let z = args
        .next()
        .expect("Missing Argument z")
        .parse::<u64>()
        .expect("z must be a number");
    main_loop(iters, x, y, z)
}
